Element: update calls to use sdf::Errors output

marcoag opened the pull request:
Signed-off-by: Marco A. Gutierrez <marco@openrobotics.org>

# 🎉 New feature

Work towards https://github.com/gazebosim/sdformat/issues/820.

## Summary

This is an update for the class `Element` making the calls use `sdf::Errors` whenever possible to avoid unwanted console outputs.

## Test it

Using any method of the `Element` class with the `errors` parameter should not print any error.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

azeey commented on include/sdf/Element.hh:
```diff
@@ -810,27 +1011,42 @@ namespace sdf
+  {
+    sdf::Errors errors;
+    std::pair<T, bool> result = this->Get<T>(errors, _key, _defaultValue);
+    for(auto& error : errors)
+    {
+        sdferr << error.Message();
+    }
```
Use `throwOrPrintErrors`

marcoag replied:
As per IM I moved `throwOrPrintErrors` into Error.cc. I opened a new PR with this changes [here](https://github.com/gazebosim/sdformat/pull/1220) since it seems like this changes and further discussion can have it's own scope.

Please review there so it can be used here once merged.

marcoag replied:
Since https://github.com/gazebosim/sdformat/pull/1220 has been merged, `ThrowOrPrintErrors` usage was added in https://github.com/gazebosim/sdformat/pull/1141/commits/8d9f1516267e36903b73e9aed7bfee371b45b1a6.

azeey commented on include/sdf/Element.hh:
```diff
@@ -848,10 +1064,23 @@ namespace sdf
+  {
+    sdf::Errors errors;
+    bool result = this->Set<T>(_value, errors);
+    for(auto& error : errors)
+    {
+        sdferr << error.Message();
+    }
```
Use `throwOrPrintErrors`

marcoag replied:
Added in https://github.com/gazebosim/sdformat/pull/1141/commits/8d9f1516267e36903b73e9aed7bfee371b45b1a6.

azeey commented on src/Element.cc:
```diff
@@ -697,22 +744,54 @@ void Element::PrintValues(const std::string &_prefix,
                           const PrintConfig &_config) const
+{
+  std::ostringstream ss;
+  sdf::Errors errors;
```
nit: This could be implemented using `PrintValues` that takes an `_errors` instead of `PrintValuesImpl`.

marcoag replied:
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

azeey commented on src/Element.cc:
```diff
@@ -697,22 +744,54 @@ void Element::PrintValues(const std::string &_prefix,
+{
+  sdf::Errors errors;
+  std::ostringstream out;
+  this->ToString(out, errors, _prefix, true, false, _config);
```
nit: This could be implemented using the `ToString` that takes an `_errors`.

marcoag replied:
Changed in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

azeey reviewed (commented):
Looks good, but I would like to see tests that cover some of the most frequently used APIs sugh as `Element::Get`.

marcoag commented on include/sdf/Element.hh:
```diff
@@ -194,24 +220,60 @@ namespace sdf
+    /// \param[out] _html Accumulated HTML for output.
+    /// \param[in] _spacing Amount of spacing for this element.
+    /// \param[out] _errors Vector of errors.
+    public: void PrintDocRightPane(std::string &_html, int _spacing,
```
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

marcoag commented on include/sdf/Element.hh:
```diff
@@ -401,12 +485,32 @@ namespace sdf
+    /// \param[out] _errors Vector of errors.
+    /// \return True if the value was successfully set, false otherwise.
+    public: template<typename T>
+            bool Set(const T &_value, sdf::Errors &_errors);
```
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

marcoag commented on include/sdf/Element.hh:
```diff
@@ -661,28 +825,32 @@ namespace sdf
     /// \param[in] _config Configuration for printing values.
     /// \param[out] _out the std::ostringstream to write output to.
+    /// \param[out] _errors Vector of errors.
     private: void PrintValuesImpl(const std::string &_prefix,
```
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

marcoag commented on include/sdf/Element.hh:
```diff
@@ -782,8 +950,29 @@ namespace sdf
+    /// \param[in] _config Configuration for printing attributes.
+    /// \param[out] _out the std::ostringstream to write output to.
+    /// \param[out] _errors Vector of errors.
+    public: void PrintAttributes(bool _includeDefaultAttributes,
```
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/83e7072e9da47fb738772bc13a998f1bca60928d.

marcoag commented:
Added tests for `Get` and updated the `Set` existing ones in https://github.com/gazebosim/sdformat/pull/1141/commits/4c83cdc9aa24303c0170c1b8eb5e294f1bf6bf9a.

The rest of functions changed here seem to only fail when the `Element` is malformed which I find tricky to create without accessing the private attributes of the class. I modified some of the current tests to make sure they don't return errors nor print any. I guess when the switch is done this should be done for all functions.

marcoag commented on include/sdf/Element.hh:
```diff
@@ -661,28 +825,32 @@ namespace sdf
     /// \param[in] _config Configuration for converting to string.
-    /// \param[out] _out the std::ostreamstream to write output to.
-    private: void ToString(const std::string &_prefix,
+    private: void ToString(std::ostringstream &_out,
```
Done in https://github.com/gazebosim/sdformat/pull/1141/commits/46f69242af581628a157356dbc13a99e3c7c6ae8.

azeey commented:
> The rest of functions changed here seem to only fail when the `Element` is malformed which I find tricky to create without accessing the private attributes of the class. I modified some of the current tests to make sure they don't return errors nor print any. I guess when the switch is done this should be done for all functions.

Yes, some of the functions are very unlikely to emit an error.

azeey commented:
> > The rest of functions changed here seem to only fail when the `Element` is malformed which I find tricky to create without accessing the private attributes of the class. I modified some of the current tests to make sure they don't return errors nor print any. I guess when the switch is done this should be done for all functions.
> 
> Yes, some of the functions are very unlikely to emit an error, but we might be able to induce some errors by providing ill formed description files. I'll try to look into this.


@marcoag I've pushed some changes to [`azeey/test_element_error_propagation`](https://github.com/gazebosim/sdformat/pull/new/azeey/test_element_error_propagation
). Namely, I added some tests that check the propagation `sdf::Errors` in some of the overloads added in this PR. 

I was unable to find a way to cause `CountNamedElements` to emit an error, but in the process of attempting to do that, I realized that by removing `SDF_ASSERT` calls, we've made it possible for users to end up with a segfault instead of an exception. This is because as opposed to throwing an exception when there is a critical problem, we now insert a fatal error into an `sdf::Errors` object  and continue executing the function. We should instead be checking if a fatal error occurred and returning early.  So in the mentioned branch, I have also added a convenience function to check if there are fatal errors after calling a function
https://github.com/gazebosim/sdformat/blob/6af4be6b2dbd13271ef88b2088ac91cea695dfa6/src/Element.cc#L1243-L1246

Would you be able to look through `Element.cc` and probably also `Param.cc` and add similar logic to places where returning early is necessary?

azeey commented:
Per video call, it would be best to revert all the changes we made code that calls `SDF_ASSERT` so that exceptions don't cause segfaults. If we remove the console printing in `sdf::Exception` (https://github.com/gazebosim/sdformat/blob/6af4be6b2dbd13271ef88b2088ac91cea695dfa6/src/Exception.cc#L52), exceptions could be handled by the end user with no errors printed to the console. I think the original goal of #820 was to remove console messages, not eliminating exceptions, but I think we started removing exceptions as well maybe hoping that it would be fairly straightforward. We should still try to remove exceptions from libsdformat, but we should leave that for another time. /cc @scpeters.

azeey commented:
Can you resolve the conflicts?

azeey reviewed (approved):
LGTM! Just a few minor comments

marcoag commented on src/Element.cc:
```diff
@@ -652,17 +677,30 @@ void ElementPrivate::PrintAttributes(bool _includeDefaultAttributes,
+{
+  std::ostringstream ss;
+  sdf::Errors errors;
+  PrintValuesImpl(errors, _prefix, true, false, _config, ss);
```
True, changed to call to `PrintValues`.

marcoag commented on src/Element.cc:
```diff
@@ -685,34 +740,66 @@ void Element::PrintValues(const std::string &_prefix,
                               const PrintConfig &_config) const
 {
+  sdf::Errors errors;
+  std::string out;
```
I changed it myself instead of committing the suggestion because the web commit suggestion thing always gave me troubles.

marcoag commented on src/Element_TEST.cc:
```diff
@@ -745,19 +765,46 @@ TEST(Element, DocRightPane)
+
+  elem->AddValue("int", "0", true, errors, "value");
+  ASSERT_EQ(errors.size(), 0u);
+  ASSERT_TRUE(elem->Set<std::string>(errors, ""));
```
Great catch! Indeed this could be a bug fix that changes behavior as it seems it's been like that before. Maybe we should log this change separately on the change log? is it easier with a separate PR for this?

I went ahead and fixed it here for now.

marcoag replied:
Since this is a bug in the code, as per video call, I opened a separate PR to keep better track of this fix [here](https://github.com/gazebosim/sdformat/pull/1256).

azeey replied:
Sounds good. Thanks!

marcoag commented on test/integration/error_output.cc:
```diff
@@ -233,6 +233,29 @@ TEST(ErrorOutput, ElementErrorOutput)
+  elem->Get<int>(errors, "key");
+  ASSERT_EQ(errors.size(), 1u);
+  EXPECT_NE(std::string::npos, errors[0].Message().find(
+      "Invalid argument. Unable to set value [test] for key[key]."));
```
So the reason for this seems a bit complicated: basically when the key exists but the type does not match, the call to the `Element->Get()` function ends up in the `Parm->Get()` [here](https://github.com/gazebosim/sdformat/blob/4464fbd413ff568f9a1c635f71fbaa4fae6b63e5/include/sdf/Param.hh#L729) where it tries to get the not-matching type on the `Get` call and the current value of the `param` and tries to set it with the new type that was asked for in the initial `Get` function. I don't know what's the use case for this, maybe because the type could have been wrongly parsed initially? 

IMHO if this is a wrong behavior and it is to be changed it probably deserves it's own issue, for now I'll add a comment with a clarification on the test.

marcoag replied:
As per video call, I opened a separated issue [here](https://github.com/gazebosim/sdformat/issues/1257) so we can track if we either modify the error message of `ValueFromStringImpl` into something more generic or we somehow modify the behavior of [the Get method](https://github.com/gazebosim/sdformat/blob/4464fbd413ff568f9a1c635f71fbaa4fae6b63e5/include/sdf/Param.hh#L712).

azeey reviewed (approved):
Thanks for iterating! I'll go ahead and merge!
